File: mayi/src/main/java/com/thanosfisherman/mayi/listeners/IPermissionBuilder.java
package com.thanosfisherman.mayi.listeners;

import com.thanosfisherman.mayi.listeners.multi.PermissionResultMultiListener;
import com.thanosfisherman.mayi.listeners.multi.RationaleMultiListener;
import com.thanosfisherman.mayi.listeners.single.PermissionResultSingleListener;
import com.thanosfisherman.mayi.listeners.single.RationaleSingleListener;

public interface IPermissionBuilder
{
    IPermissionBuilder onErrorListener(MayiErrorListener errorListener);
    void check();

    interface Permission
    {
        SinglePermissionBuilder withPermission(String permission);

        IPermissionBuilder.MultiPermissionBuilder withPermissions(String... permissions);

        //IPermissionBuilder.MultiPermissionBuilder withPermissions(Collection<String> permissions);
    }

    interface SinglePermissionBuilder extends IPermissionBuilder
    {
        SinglePermissionBuilder onResult(PermissionResultSingleListener response);
        SinglePermissionBuilder onRationale(RationaleSingleListener rationale);
    }

    interface MultiPermissionBuilder extends IPermissionBuilder
    {
        MultiPermissionBuilder onResult(PermissionResultMultiListener response);
        MultiPermissionBuilder onRationale(RationaleMultiListener rationale);
    }
}

File: mayi/src/main/java/com/thanosfisherman/mayi/listeners/IPermissionBuilder.kt
package com.thanosfisherman.mayi.listeners

import com.thanosfisherman.mayi.listeners.multi.PermissionResultMultiListener
import com.thanosfisherman.mayi.listeners.multi.RationaleMultiListener
import com.thanosfisherman.mayi.listeners.single.PermissionResultSingleListener
import com.thanosfisherman.mayi.listeners.single.RationaleSingleListener

interface IPermissionBuilder {
    fun onErrorListener(errorListener: MayiErrorListener): IPermissionBuilder
    fun check()

    interface Permission {
        fun withPermission(permission: String): SinglePermissionBuilder

        fun withPermissions(vararg permissions: String): IPermissionBuilder.MultiPermissionBuilder

        //IPermissionBuilder.MultiPermissionBuilder withPermissions(Collection<String> permissions);
    }

    interface SinglePermissionBuilder : IPermissionBuilder {
        fun onResult(response: PermissionResultSingleListener): SinglePermissionBuilder
        fun onRationale(rationale: RationaleSingleListener): SinglePermissionBuilder
    }

    interface MultiPermissionBuilder : IPermissionBuilder {
        fun onResult(response: PermissionResultMultiListener): MultiPermissionBuilder
        fun onRationale(rationale: RationaleMultiListener): MultiPermissionBuilder
    }
}
